File: c/2.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <stdbool.h>
#include <ctype.h>
#include "../../Modules/input.h"
#define defaultInput "../../Inputs/2021_2.txt"
#include "../../Modules/vector.h"


struct dataline {
    int amt;
    char *str;
};

size_t datalineSize(void *e) {
    struct dataline *dline = (struct dataline*)e;
    return sizeof(int) + strlen(dline->str);
}

void *copy_dataline(void *e, size_t size) {
    struct dataline *old = (struct dataline*)e;
    struct dataline *new = (struct dataline*)calloc(1, sizeof(struct dataline));
    new->str = (char*)calloc(strlen(old->str) + 1, sizeof(char));
    strcpy(new->str, old->str);
    new->amt = old->amt;

    return new;
}

int part1(char *fileName) {
    struct vector *input_data = multiLine(fileName);
    struct vector *data = createVector(datalineSize, copy_dataline);

    for (int i = 0; i < input_data->len; i++) {
        struct dataline *dline = (struct dataline*)calloc(1, sizeof(struct dataline));
        char *line = (char*)input_data->arr[i];
        char *p = strtok(line, " ");
        dline->str = (char*)calloc(strlen(p), sizeof(char));
        strcpy(dline->str, p);
        p = strtok(NULL, " ");
        dline->amt = atoi(p);
        appendVector(data, dline);
    }

    int x = 0, y = 0;
    for (int i = 0; i < data->len; i++) {
        struct dataline *dline = (struct dataline*)data->arr[i];
        if (strcmp(dline->str, "forward") == 0) {
            x += dline->amt;
        } else {
            y += (strcmp(dline->str, "up") == 0) ? -dline->amt : dline->amt;
        }
    }

    return x * y;
}

int part2(char *fileName) {
    struct vector *input_data = multiLine(fileName);
    struct vector *data = createVector(datalineSize, copy_dataline);

    for (int i = 0; i < input_data->len; i++) {
        struct dataline *dline = (struct dataline*)calloc(1, sizeof(struct dataline));
        char *line = (char*)input_data->arr[i];
        char *p = strtok(line, " ");
        dline->str = (char*)calloc(strlen(p), sizeof(char));
        strcpy(dline->str, p);
        p = strtok(NULL, " ");
        dline->amt = atoi(p);
        appendVector(data, dline);
    }


    int aim = 0, x = 0, y = 0;
    for (int i = 0; i < data->len; i++) {
        struct dataline *dline = (struct dataline*)data->arr[i];
        if (strcmp(dline->str, "forward") == 0) {
            x += dline->amt;
            y += aim * dline->amt;
        } else {
            aim += (strcmp(dline->str, "up") == 0) ? -dline->amt : dline->amt;
        }
    }

    return x * y;
}

int main (int argc, char *argv[]) {
    char *inputPath = defaultInput;
    if (argc > 1) {
        inputPath = argv[1];
    }

    clock_t t;
    t = clock(); 
    int p1 = part1(inputPath);
    t = clock() - t; 
    double t_p1 = ((double)t) / CLOCKS_PER_SEC;
    printf("\nPart 1:\nAnswer: %d\nRan in %f seconds\n", p1, t_p1);

    t = clock(); 
    int p2 = part2(inputPath);
    t = clock() - t;
    double t_p2 = ((double)t) / CLOCKS_PER_SEC;
    printf("\nPart 2:\nAnswer: %d\nRan in %f seconds\n", p2, t_p2);


    return 0;
}

File: python/2.py
import re
import sys
from pathlib import Path
from typing import Any, List, Optional, Tuple

sys.path.append(str(Path(__file__).parent.parent.parent))
from Modules.timer import Timer
def part1(data):
    """ 2021 Day 2 Part 1

    >>> part1(['forward 5', 'down 5', 'forward 8', 'up 3', 'down 8', 'forward 2'])
    150
    """

    # 0 is horizontal position, 1 is depth
    pos = [0] * 2

    for line in data:
        split = line.split(' ')
        num = int(split[-1])
        if line[0] == 'f':
            pos[0] += num
        elif line[0] == 'd':
            pos[1] += num
        elif line[0] == 'u':
            pos[1] -= num

    product = 1
    for i in pos:
        product *= i

    return product


def part2(data):
    """ 2021 Day 2 Part 2

    >>> part2(['forward 5', 'down 5', 'forward 8', 'up 3', 'down 8', 'forward 2'])
    900
    """

    pos = [0] * 2
    aim = 0
    for line in data:
        split = line.split(' ')
        num = int(split[-1])
        if line[0] == 'f':
            pos[0] += num
            pos[1] += num * aim
        elif line[0] == 'd':
            aim += num
        elif line[0] == 'u':
            aim -= num

    product = 1
    for i in pos:
        product *= i

    return product


def main(input_path: Optional[Path | str]=None, verbose: bool=False) -> Tuple[Tuple[Any, float]]:
    if not input_path:
        if not (input_path := sys.argv[1] if len(sys.argv) > 1 else None):
            year, day = re.findall(r'\d+', str(__file__))[-2:]
            input_path = Path(Path(__file__).parent.parent.parent, "Inputs", f"{year}_{day}.txt")
    
    with open(input_path, encoding='UTF-8') as f:
        data = [line.strip('\n') for line in f.readlines()]

    with Timer() as p1_time:
        p1 = part1(data)

    if verbose:
        print(f"\nPart 1:\nProduct: {p1}\nRan in {p1_time.elapsed:0.4f} seconds")

    with Timer() as p2_time:
        p2 = part2(data)

    if verbose:
        print(f"\nPart 2:\nProduct: {p2}\nRan in {p2_time.elapsed:0.4f} seconds")

    return [(p1, p1_time.elapsed), (p2, p2_time.elapsed)]


if __name__ == "__main__":
    main(verbose=True)